pkg/cluster/clustermsi.go

```diff
@@ -0,0 +1,172 @@
+		return err
+	}
+
+	clusterMsiResourceId, err := m.clusterMsiResourceId()
+	if err != nil {
+		return err
+	}
+
+	uaMsiRequest := dataplane.UserAssignedMSIRequest{
+		IdentityURL: m.doc.OpenShiftCluster.Identity.IdentityURL,
+		ResourceIDs: []string{clusterMsiResourceId.String()},
+		TenantID:    m.doc.OpenShiftCluster.Identity.TenantID,
+	}
+
+	msiCredObj, err := m.msiDataplane.GetUserAssignedIdentities(ctx, uaMsiRequest)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	expirationDate := now.AddDate(0, 0, msiCertValidityDays)
+	secretProperties := store.SecretProperties{
+		Enabled:   true,
+		Expires:   expirationDate,
+		Name:      secretName,
+		NotBefore: now,
+	}
+
+	return m.clusterMsiKeyVaultStore.SetCredentialsObject(ctx, secretProperties, msiCredObj.CredentialsObject)
+}
+
+// initializeClusterMsiClients intializes any Azure clients that use the cluster
+// MSI certificate.
+func (m *manager) initializeClusterMsiClients(ctx context.Context) error {
+	secretName, err := m.clusterMsiSecretName()
+	if err != nil {
+		return err
+	}
+
+	kvSecret, err := m.clusterMsiKeyVaultStore.GetCredentialsObject(ctx, secretName)
+	if err != nil {
+		return err
+	}
+
+	// This code assumes that there is only one cluster MSI and will have to be
+	// refactored if we ever use more than one.
+	if kvSecret.CredentialsObject.ExplicitIdentities == nil {
+		return errors.New("found nil ExplicitIdentities in cluster MSI CredentialsObject")
+	} else if len(kvSecret.CredentialsObject.ExplicitIdentities) == 0 {
+		return errors.New("found empty ExplicitIdentities in cluster MSI CredentialsObject")
+	} else if len(kvSecret.CredentialsObject.ExplicitIdentities) > 1 {
+		m.log.Warning("unexpectedly found more than one entry in ExplicitIdentities in cluster MSI CredentialsObject; will attempt to instantiate a cluster MSI credential using the first one")
+	}
+
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].ClientID == nil {
+		return errors.New("found nil ClientID while parsing cluster MSI CredentialsObject")
+	}
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].TenantID == nil {
+		return errors.New("found nil TenantID while parsing cluster MSI CredentialsObject")
+	}
+	if kvSecret.CredentialsObject.ExplicitIdentities[0].ClientSecret == nil {
+		return errors.New("found nil ClientSecret while parsing cluster MSI CredentialsObject")
+	}
+
+	clientId := *kvSecret.CredentialsObject.ExplicitIdentities[0].ClientID
+	tenantId := *kvSecret.CredentialsObject.ExplicitIdentities[0].TenantID
+	certData := *kvSecret.CredentialsObject.ExplicitIdentities[0].ClientSecret
+
+	decodedCertData, err := base64.StdEncoding.DecodeString(certData)
+	if err != nil {
+		return err
+	}
+
+	certs, key, err := azidentity.ParseCertificates(decodedCertData, nil)
+	if err != nil {
+		return err
+	}
+
+	cred, err := azidentity.NewClientCertificateCredential(tenantId, clientId, certs, key, m.env.Environment().ClientCertificateCredentialOptions())
```

Comment:
You can just use this guy: https://github.com/Azure/msi-dataplane/blob/94e135accf5ed734b30612be73fdb8537d8ae2ce/pkg/dataplane/identity.go#L43

Reply:
I tried using it, but I got stuck :/

The key vault storage client only accepts the `CredentialsObject` type, so after getting it from the key vault, I have to make the `UserAssignedIdentities` object myself. But there's no constructor function and the `cloud` field is package-private, so I can't set it. This won't work in FairFax.

Am I understanding correctly? If so, I think the dataplane module either needs a `UserAssignedIdentities` constructor function or some other tweak.

Reply:
I'll fix this in `msi-dataplane` and will post an update.

Reply:
PR to fix this: https://github.com/Azure/msi-dataplane/pull/20/files

Reply:
Yep, it worked well this time! All fixed.